libxbot-service/src/Scheduler.cpp

```diff
@@ -0,0 +1,77 @@
+#include <xbot-service/Lock.hpp>
+#include <xbot-service/Scheduler.hpp>
+#include <xbot-service/portable/system.hpp>
+
+xbot::service::Scheduler::~Scheduler() {
+  thread::deinitialize(thread_);
+  mutex::deinitialize(&state_mutex_);
+}
+
+bool xbot::service::Scheduler::Start() {
+  stopped_ = false;
+
+  if (!mutex::initialize(&state_mutex_)) {
+    return false;
+  }
+
+#ifdef XBOT_ENABLE_STATIC_STACK
+  if (!thread::initialize(thread_, ThreadLoopHelper, this, thread_stack_, sizeof(thread_stack_), "Scheduler"))
+#else
+  if (!thread::initialize(thread_, ThreadLoopHelper, this, nullptr, 0, "Scheduler"))
+#endif  // XBOT_ENABLE_STATIC_STACK
+  {
+    return false;
+  }
+
+  return true;
+}
+
+void xbot::service::Scheduler::ThreadLoop() {
+  while (true) {
+    {
+      Lock lk(&state_mutex_);
+      if (stopped_) return;
+    }
+
+    uint32_t now_micros = system::getTimeMicros();
+
+    // TODO: Handle overflow
+
+    if (schedules_head_->next_execution_ > now_micros) {
+      uint32_t sleep_time = schedules_head_->next_execution_ - now_micros;
+      // TODO: Sleep in an interruptible way (new schedule, stopped_ set).
+      //       chThdSleepUntil() might be a candidate, but it's not interruptible.
+      now_micros = system::getTimeMicros();
+      (void)sleep_time;
+    }
+
+    // TODO: Locking.
+    for (Schedule* schedule = schedules_head_; schedule != nullptr; schedule = schedule->next_) {
+      if (now_micros >= schedule->next_execution_) {
+        schedule->callback_();
+        schedule->next_execution_ = now_micros + schedule->interval_;
+        // Move schedule to the correct position in the list, so that the list is sorted by next_execution
```

Comment:
Is sorting even worth here from a runtime perspective?

Assume not sorted and while scanning you store the min(time_to_next_schedule), then we would have linear time (single sweep, execute all which are pending, get the min sleep time). So that's the one to beat.

When having a sorted list, we could cancel early because if the next callback is not due, all callbacks after that are not due as well. But then we need to insertion sort the list which requires partly scanning the list.

the question now is, is this usually better or worse than a single sweep. I'm not sure, I feel that its only worth doing for the scenario where there's many callbacks but each one gets rarely called, so iterating the list would basically a waste of time.